artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/SimpleAddressManager.java

```diff
@@ -116,7 +121,15 @@ public Binding removeBinding(final SimpleString uniqueName, Transaction tx) thro
          return null;
       }
 
-      removeBindingInternal(binding.getA().getAddress(), uniqueName);
+      SimpleString address = binding.getA().getAddress();
+      removeBindingInternal(address, uniqueName);
+      Collection<Binding> directBindings = directBindingMap.get(address);
+      if (directBindings != null) {
+         directBindings.remove(binding.getA());
+         if (directBindings.isEmpty()) {
```

Comment:
Isn't there a race condition here in that the directBindings collection might report empty here but a new binding could be added between this check and the next call to remove the bindings entry from the Map meaning that the tracked state is now inconsistent with reality?

Reply:
Should we perhaps put this in a `ConcurrentHashMap#compute` lambda instead?

Reply:
I have now used `compute` which should take advantage of the striped locking nature of the map, and should remove the race condition mentioned.